ledger/voters.go

```diff
@@ -0,0 +1,351 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package ledger
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/algorand/go-deadlock"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/crypto/compactcert"
+	"github.com/algorand/go-algorand/crypto/merklearray"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+)
+
+// The votersTracker maintains the Merkle tree for the most recent
+// commitments to online accounts for compact certificates.
+//
+// We maintain multiple Merkle trees: we might commit to a new Merkle tree in
+// block X, but we need the Merkle tree from block X-params.CompactCertBlocks
+// to build the compact certificate for block X.
+//
+// votersTracker is kind-of like a tracker, but hangs off the acctupdates
+// rather than a direct ledger tracker.  We don't have an explicit interface
+// for such an "accounts tracker" yet, however.
+type votersTracker struct {
+	// round contains the top online accounts in a given round.
+	//
+	// To avoid increasing block latency, we include a Merkle commitment
+	// to the top online accounts as of block X in the block header of
+	// block X+CompactCertVotersLookback.  This gives each node some time
+	// to construct this Merkle tree, before its root is needed in a block.
+	//
+	// This round map is indexed by the block X, using the terminology from
+	// the above example, to be used in X+CompactCertVotersLookback.
+	//
+	// We maintain round entries for two reasons:
+	//
+	// The first is to maintain the tree for an upcoming block -- that is,
+	// if X+Loookback<Latest.  The block evaluator can ask for the root of
+	// the tree to propose and validate a block.
+	//
+	// The second is to construct compact certificates.  Compact certificates
+	// are formed for blocks that are a multiple of CompactCertRounds, using
+	// the Merkle commitment to online accounts from the previous such block.
+	// Thus, we maintain X in the round map until we form a compact certificate
+	// for round X+CompactCertVotersLookback+CompactCertRounds.
```

Comment:
It is indeed the case that `CompactCertVotersLookback` is on the order of 16, and `CompactCertRounds` is probably something like 128 or 1024.

However, I believe the comment is correct: the two intervals have to be added together.  We need the compact cert to certify the next block that is a multiple of `CompactCertRounds`, because that block will contain the commitment to the next group of voters (which will enable verifying the next compact cert, etc).  And the `CompactCertVotersLookback` addition is needed to give the ledger time to compute the Merkle tree, in case there are many voters.